lib/srv/db/server.go

```diff
@@ -655,19 +655,21 @@ func (s *Server) getServerInfoFunc(database types.Database) func() (types.Resour
 // getServerInfo returns up-to-date database resource e.g. with updated dynamic
 // labels.
 func (s *Server) getServerInfo(database types.Database) (types.Resource, error) {
+	s.mu.Lock()
+	if s.cfg.CloudLabels != nil {
+		s.cfg.CloudLabels.Apply(database)
+	}
```

Comment:
Does this Apply approach in `getServerInfo` will work properly with dynamic resource where db resources are matched against each others to detect changes. If the `CloudLabels.Apply(` is applies only to one resource the reconciler will always treat this resource as new. 

Could you adda UT for dynamic resource flow ?

Reply:
If I got this correctly, I don't think there will be any difference in change detection, as it is done with the database's copy (which, before this PR, already got the cloud labels applied).

> Does this Apply approach in getServerInfo will work properly with dynamic resource where db resources are matched against each others to detect changes.

Isn't this the correct behavior? If the cloud labels are updated, the database should be announced so the auth server can edit its contents, right?

> Could you adda UT for dynamic resource flow ?

Added!

Reply:
I'm not sure that this works correctly:

The `Stopped database` `Heartbeat exited` and ` Starting Database heartbeat with ` calls are invoked  triggered even if db was not changed: 
```
2023-07-11T12:27:15+02:00 INFO [DB:SERVIC] db rdsproxy-marek-sqlsever01 updated, updating. services/reconciler.go:169
2023-07-11T12:27:15+02:00 DEBU [DB:SERVIC] Stopped database "rdsproxy-marek-sqlsever01". db/server.go:462
2023-07-11T12:27:15+02:00 DEBU [DB:SERVIC] Heartbeat exited. srv/heartbeat.go:290
2023-07-11T12:27:15+02:00 DEBU [META]      Fetched metadata for "Database(Name=rdsproxy-marek-sqlsever01, Type=rdsproxy, Labels=map[account-id:278576220453 engine:SQLSERVER region:eu-central-1 teleport.dev/origin:cloud vpc-id:vpc-59e96733])": Region:"eu-central-1" Redshift:<> RDS:<> AccountID:"278576220453" ElastiCache:<> SecretStore:<> MemoryDB:<> RDSProxy:<Name:"rdsproxy-marek-sqlsever01" ResourceID:"prx-0ce165147db5baa94" > RedshiftServerless:<> OpenSearch:<> . cloud/meta.go:109
2023-07-11T12:27:15+02:00 DEBU [DB:SERVIC] Starting Database heartbeat with announce period: 1m0s, keep-alive period 5m42.73036562s, poll period: 5s srv/heartbeat.go:164

2023-07-11T12:27:16+02:00 INFO [DB:SERVIC] db rdsproxy-marek-sqlsever01 updated, updating. services/reconciler.go:169
2023-07-11T12:27:16+02:00 DEBU [DB:SERVIC] Stopped database "rdsproxy-marek-sqlsever01". db/server.go:462
2023-07-11T12:27:16+02:00 DEBU [DB:SERVIC] Heartbeat exited. srv/heartbeat.go:290
2023-07-11T12:27:16+02:00 DEBU [META]      Fetched metadata for "Database(Name=rdsproxy-marek-sqlsever01, Type=rdsproxy, Labels=map[account-id:278576220453 engine:SQLSERVER region:eu-central-1 teleport.dev/origin:cloud vpc-id:vpc-59e96733])": Region:"eu-central-1" Redshift:<> RDS:<> AccountID:"278576220453" ElastiCache:<> SecretStore:<> MemoryDB:<> RDSProxy:<Name:"rdsproxy-marek-sqlsever01" ResourceID:"prx-0ce165147db5baa94" > RedshiftServerless:<> OpenSearch:<> . cloud/meta.go:109
2023-07-11T12:27:16+02:00 DEBU [DB:SERVIC] Starting Database heartbeat with announce period: 1m0s, keep-alive period 5m46.235836108s, poll period: 5s srv/heartbeat.go:164



2023-07-11T12:27:21+02:00 INFO [DB:SERVIC] db rdsproxy-marek-sqlsever01 updated, updating. services/reconciler.go:169
2023-07-11T12:27:21+02:00 DEBU [DB:SERVIC] Stopped database "rdsproxy-marek-sqlsever01". db/server.go:462
2023-07-11T12:27:21+02:00 DEBU [DB:SERVIC] Heartbeat exited. srv/heartbeat.go:290
2023-07-11T12:27:21+02:00 DEBU [META]      Fetched metadata for "Database(Name=rdsproxy-marek-sqlsever01, Type=rdsproxy, Labels=map[account-id:278576220453 engine:SQLSERVER region:eu-central-1 teleport.dev/origin:cloud vpc-id:vpc-59e96733])": Region:"eu-central-1" Redshift:<> RDS:<> AccountID:"278576220453" ElastiCache:<> SecretStore:<> MemoryDB:<> RDSProxy:<Name:"rdsproxy-marek-sqlsever01" ResourceID:"prx-0ce165147db5baa94" > RedshiftServerless:<> OpenSearch:<> . cloud/meta.go:109
2023-07-11T12:27:21+02:00 DEBU [DB:SERVIC] Starting Database heartbeat with announce period: 1m0s, keep-alive period 5m52.522710805s, poll period: 5s srv/heartbeat.go:164
2023-07-11T12:27:21+02:00 DEBU [DB:SERVIC] Started Database(Name=rdsproxy-marek-sqlsever01, Type=rdsproxy, Labels=map[account-id:278576220453 aws/my:value engine:SQLSERVER region:eu-central-1 teleport.dev/origin:cloud vpc-id:vpc-59e96733]). db/server.go:452


2023-07-11T12:27:41+02:00 INFO [DB:SERVIC] db rdsproxy-marek-sqlsever01 updated, updating. services/reconciler.go:169
2023-07-11T12:27:41+02:00 DEBU [DB:SERVIC] Stopped database "rdsproxy-marek-sqlsever01". db/server.go:462
2023-07-11T12:27:41+02:00 DEBU [DB:SERVIC] Heartbeat exited. srv/heartbeat.go:290
2023-07-11T12:27:41+02:00 DEBU [META]      Fetched metadata for "Database(Name=rdsproxy-marek-sqlsever01, Type=rdsproxy, Labels=map[account-id:278576220453 engine:SQLSERVER region:eu-central-1 teleport.dev/origin:cloud vpc-id:vpc-59e96733])": Region:"eu-central-1" Redshift:<> RDS:<> AccountID:"278576220453" ElastiCache:<> SecretStore:<> MemoryDB:<> RDSProxy:<Name:"rdsproxy-marek-sqlsever01" ResourceID:"prx-0ce165147db5baa94" > RedshiftServerless:<> OpenSearch:<> . cloud/meta.go:109
2023-07-11T12:27:41+02:00 DEBU [DB:SERVIC] Starting Database heartbeat with announce period: 1m0s, keep-alive period 5m15.903590416s, poll period: 5s srv/heartbeat.go:164



2023-07-11T12:28:01+02:00 INFO [DB:SERVIC] db rdsproxy-marek-sqlsever01 updated, updating. services/reconciler.go:169
2023-07-11T12:28:01+02:00 DEBU [DB:SERVIC] Stopped database "rdsproxy-marek-sqlsever01". db/server.go:462
2023-07-11T12:28:01+02:00 DEBU [DB:SERVIC] Heartbeat exited. srv/heartbeat.go:290
2023-07-11T12:28:01+02:00 DEBU [META]      Fetched metadata for "Database(Name=rdsproxy-marek-sqlsever01, Type=rdsproxy, Labels=map[account-id:278576220453 engine:SQLSERVER region:eu-central-1 teleport.dev/origin:cloud vpc-id:vpc-59e96733])": Region:"eu-central-1" Redshift:<> RDS:<> AccountID:"278576220453" ElastiCache:<> SecretStore:<> MemoryDB:<> RDSProxy:<Name:"rdsproxy-marek-sqlsever01" ResourceID:"prx-0ce165147db5baa94" > RedshiftServerless:<> OpenSearch:<> . cloud/meta.go:109
2023-07-11T12:28:01+02:00 DEBU [DB:SERVIC] Starting Database heartbeat with announce period: 1m0s, keep-alive period 5m49.998280793s, poll period: 5s srv/heartbeat.go:164

```

The [`CompareResources(new, registered) != Equal`](https://github.com/gravitational/teleport/blob/0ee91f6c37eb3a8c5bc98b15f857c95902df50b2/lib/services/reconciler.go#L108-L107 ) reconciler logic will always be evaluated to `TRUE` because after adding CloudLabels to the current resource `registered` the `new` resource doesn't contain CloudLabels and object differs. 

In the result in each iteration of cloud watchers a cloud database with CloudLables will be treated as updated and the 
[updateDatabase](https://github.com/gravitational/teleport/blob/02058a18c8267db78e0a2ce068a39bef7fc8549d/lib/srv/db/server.go#L535) callback will be called.  We don't want to do that because [updateDatabase](https://github.com/gravitational/teleport/blob/02058a18c8267db78e0a2ce068a39bef7fc8549d/lib/srv/db/server.go#L535) will stop database and deregister db from backend and add updated resource as a new.